### PR TITLE
Provide completions for s3cmd

### PR DESCRIPTION
**Pull request by mqudsi**
No longer auto-generated. Everything has been summarized. Supressing
file completions for initial command, providing list of valid initial
commands, filtering --options by subcommand.

### CODE REVIEW

File: share/completions/s3cmd.fish [hunk 1]
```diff
@@ -0,0 +1,178 @@
+# s3cmd
+
+function __s3cmd_is_remote_path
+	commandline -pct | string match -r -- "^s3://.*"
```
Comment (faho) [inline share/completions/s3cmd.fish]: You'll usually want to use the "-q" flag for `string match`.

It's okay here since this is only used as a complete condition, which has its output already silenced.

File: share/completions/s3cmd.fish [hunk 2]
```diff
@@ -0,0 +1,178 @@
...
+end
+
+function __s3cmd_is_valid_remote_path
+	commandline -pct | string match -r -- "^s3://.*?/.*"
+end
+
+function __prev_tokens
+	set __tokens (commandline -pco)
+	for arg in $argv
+		if echo $__tokens[-1] | string match $arg
```
Comment (faho) [inline share/completions/s3cmd.fish]: `string match -q -- $argv $__tokens[-1]` - the `echo` is unnecessary.

File: share/completions/s3cmd.fish [hunk 3]
```diff
@@ -0,0 +1,178 @@
...
+			return 0
+		end
+	end
+	return 1
+end
+
+function __initial_command
+	set __tokens (commandline -pco)
+	string match (count $__tokens) 1
```
Comment (faho) [inline share/completions/s3cmd.fish]: This is a bit unorthodox.

Strictly, this would be `set -q __tokens[1]; and not set -q __tokens[2]`. You could probably get by without the first check, depending on what exactly you want to do with this.

Alternatively, `test (count $__tokens) = 1`.

Comment (faho) [inline share/completions/s3cmd.fish]: There is also `__fish_is_first_token` already. That one skips options.

Personally, for subcommands, I prefer `not __fish_seen_subcommand_from $cmds` (see e.g. the systemctl completions). That allows for options and option-arguments, as long as they don't look like a command.

For a completely correct solution, you're going to need to do something like the git completions, or use argparse (which I'm converting both the git and systemctl completions to).

File: share/completions/s3cmd.fish (hunk 2 above)
Comment (faho) [inline share/completions/s3cmd.fish]: Also, this is basically `__fish_seen_subcommand_from`.

**Comment by mqudsi** [comment]
@faho thanks for reviewing; I knew there would be a significant amount of feedback if I opened this a PR rather than merging directly despite the fact that it works.

I had a hard time finding a good reference on the best practices for writing completions and each completion file I looked at was different than the other (though of course I didn't go through too many). Perhaps we can add a more-indepth article about writing completions to the online resources, as the existing seemed to only cover the very basic usage of `complete`.

I didn't look at the completions for sysctl, I tried to base mine off of `apt-get` but found it too hard to follow and so ended up reinventing the wheel :)

It doesn't seem like it should be too hard to take these suggestions into account (I'm going to hold off on `argparse` at this time, if it's OK with you); let me review and push.

**Comment by faho** [comment]
>I had a hard time finding a good reference on the best practices for writing completions and each completion file I looked at was different than the other

Yes, they are. They were written at different times by different people. Obviously I prefer my own style.

>Perhaps we can add a more-indepth article about writing completions to the online resources

That would be nice to have, yes.

>I didn't look at the completions for sysctl

Note: `systemctl`, which is basically systemd's main command. `sysctl` is that thing for tweaking kernel knobs.

The completions for it are nice (except for the use of `eval`), and easier to follow than the massive git script.

>I'm going to hold off on argparse at this time, if it's OK with you

Sure.

>let me review and push.

Yes. You can push directly to Integration_2.7.0 if you want - no need to review again, unless you prefer that.

**Comment by mqudsi** [comment]
OK, thanks again.
